### PR TITLE
Add sum and mean to Num::Grad::Variable

### LINKED ISSUE #84: `sum` and `mean` not defined for Num::Grad::Variable
I am continuing to work on toy models using `num.cr` while following some PyTorch examples. As I was setting up the forward calculations using a gradient variable, I needed `sum` and `mean` functions which are currently not implemented.

### PR DESCRIPTION
**Pull request by nogginly**
Fixes #84

@christopherzimmerman, this PR adds `sum` and `mean` to `Num::Grad::Variable` with backprop using their derivatives. 
* Both CPU and OCL are supported.
* Test cases are included

I also added `Num.as_tensor(val, like: Tensor)` helper to convert a single scalar into a Tensor matching the type. I thought about adding a `Num::Grad::Context#variable` method variant but there was already one that took a `Number` and I didn't want to mess with that and break any compatibility. Let me know if there's a different convention you'd prefer I follow.
* You can see the the problem I ran into in `Num::Grad::Variable#mean` where I wanted to make a tensor with the size of another tensor and didn't want to make a different `#mean` for CPU and OCL when this lets Crystal sort things out.

### CODE REVIEW

File: src/grad/variable.cr [hunk 1]
```diff
@@ -205,6 +205,39 @@ class Num::Grad::Variable(T)
...
+  #
+  # ## Examples
+  #
+  # ```
+  # ctx = Num::Grad::Context(Tensor(Float64, CPU(Float64))).new
+  # x = ctx.variable([1.0, 2.0, 3.0, 4.0])
+  # x.sum # => [10.0]
+  # ```
+  def sum : Num::Grad::Variable(T)
+    result = @context.variable(Num.sum(@value, axis: -1, dims: true))
```
Comment (christopherzimmerman) [inline src/grad/variable.cr:218]: This method signature isn't quite accurate for what is being done here.  This is reducing along the final axis rather than reducing to a scalar, which is going to cause unexpected behavior in your `mean` method for higher dimensions since it will be using the size of the entire `Tensor` rather than per axis.  If the goal is forward/backward for a total summation, `@value.sum` should be what you need (which is not currently supported by the OCL backend).

**Review by christopherzimmerman — CHANGES_REQUESTED** [review]
Changes look good overall, the math is correct, just have one comment about the type of reduction being done.  Thanks for all of these PRs, very nice to have people using the library!

File: src/grad/variable.cr (hunk 1 above)
Comment (nogginly) [reply src/grad/variable.cr:218]: Good catch. I was OK with the `#sum` reducing along the final axis; but I didn't think through the impact on `#size` when implementing `#mean`. Ok, let me see what I can do.

Comment (nogginly) [reply src/grad/variable.cr]: @christopherzimmerman, I've modified the `#sum` and `#mean` methods to accept `axis` as a parameter. And I've refactored the tests and expanded them to test for 1D and 2D tensor variables.
